### PR TITLE
Microsoft 365 Defender: Feature - Added on_poll action and configurable timeout

### PR DESCRIPTION
**Pull request by ricardom-splunk**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [ ] The manual_readme_content.md has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [x] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [x] Verify all checks are passing.
- [x] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [x] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe):

## Security Considerations (REQUIRED)
- [N] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers),
  please document them in the `manual_readme_content.md`.
- [N] If this is a new connector or you are adding new actions
    - [ ] Please document in the `manual_readme_content.md` all methods (eg, OAuth) used to authenticate
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `manual_readme_content.md`.
- [N] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [N] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [N] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
Implemented on_poll action

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.
The on_poll action was implemented not to consider any input arguments and bases the polling on default values.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

File: microsoft365defender.json [hunk 1]
```diff
@@ -45,11 +45,32 @@
...
+        },
         "non_interactive": {
             "description": "Non-Interactive Auth",
             "data_type": "boolean",
             "default": false,
-            "order": 4
+            "order": 5
+        },
+        "max_incidents_for_polling": {
+            "description": "Maximum Incidents for scheduled/ interval polling for each cycle",
```
Comment (bb-splunk) [inline microsoft365defender.json]:
```suggestion
            "description": "Maximum Incidents for scheduled/interval polling for each cycle",
```

File: microsoft365defender.json [hunk 2]
```diff
@@ -45,11 +45,32 @@
...
+            "order": 4
+        },
         "non_interactive": {
             "description": "Non-Interactive Auth",
             "data_type": "boolean",
             "default": false,
-            "order": 4
+            "order": 5
+        },
+        "max_incidents_for_polling": {
```
Comment (bb-splunk) [inline microsoft365defender.json]: Maybe `per_poll` instead of `for_polling`?

File: microsoft365defender.json [hunk 3]
```diff
@@ -45,11 +45,32 @@
...
+            "default": 1000,
+            "order": 6
+        },
+        "start_time": {
+            "description": "Start time for schedule/interval/manual poll (Use this format: 1970-01-01T00:00:00Z)",
+            "data_type": "string",
+            "order": 7
+        },
+        "filter": {
+            "description": "Filter incidents based on property",
```
Comment (bb-splunk) [inline microsoft365defender.json]: Maybe an example would be in order?

File: microsoft365defender_connector.py [hunk 4]
```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
...
+        Parameters:
+            :param action_result: object of ActionResult class
+            :param time: object of time
+        Returns:
+            :return: status(phantom.APP_SUCCESS/phantom.APP_ERROR)
+        """
+        # Check that given time must not be before 1970-01-01T00:00:00Z.
+        if time < datetime.strptime("1970-01-01T00:00:00Z", DEFENDER_APP_DT_STR_FORMAT):
+            return phantom.APP_ERROR
+        return phantom.APP_SUCCESS
```
Comment (bb-splunk) [inline microsoft365defender_connector.py]:
```suggestion
        return time >= datetime.strptime("1970-01-01T00:00:00Z", DEFENDER_APP_DT_STR_FORMAT)
```
Why not use a regular bool but a wrapper?

File: microsoft365defender_connector.py [hunk 5]
```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
...
+            max_incidents = config.get(DEFENDER_CONFIG_FIRST_RUN_MAX_INCIDENTS,
+                                       DEFENDER_INCIDENT_DEFAULT_LIMIT_FOR_SCHEDULE_POLLING)
+            ret_val, max_incidents = self._validate_integer(action_result, max_incidents, "max_incidents")
+            if phantom.is_fail(ret_val):
+                return action_result.get_status()
+
+            if self._state.get(STATE_FIRST_RUN, True):
+                self._state[STATE_FIRST_RUN] = False
+            elif last_time := self._state.get(STATE_LAST_TIME):
+                last_modified_time = last_time  # noqa: F841
```
Comment (bb-splunk) [inline microsoft365defender_connector.py]: Why is this noqa needed?

File: microsoft365defender_connector.py [hunk 6]
```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
...
+
+        start_time_filter = f"lastUpdateDateTime ge {last_modified_time}"
+        poll_filter += start_time_filter if not poll_filter else f" and {start_time_filter}"
+
+        endpoint = "{0}{1}".format(DEFENDER_MSGRAPH_API_BASE_URL, DEFENDER_LIST_INCIDENTS_ENDPOINT)
+        endpoint += "?$expand=alerts"
```
Comment (bb-splunk) [inline microsoft365defender_connector.py]:
```suggestion
        endpoint = "{0}{1}?$expand=alerts".format(DEFENDER_MSGRAPH_API_BASE_URL, DEFENDER_LIST_INCIDENTS_ENDPOINT)
```
Needs formatting though

File: microsoft365defender_connector.py [hunk 7]
```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
...
+            artifact['container_id'] = cid
+        ret_val, message, _ = self.save_artifacts(artifacts)
+
+        self.debug_print("save_artifacts returns, value: {}, reason: {}".format(ret_val, message))
+
+    @staticmethod
+    def _create_alert_artifacts(alert):
+
+        return {'label': 'alert', 'name': alert.get('title'), 'source_data_identifier': alert.get('id'),
+                'data': alert, 'cef': alert}
```
Comment (bb-splunk) [inline microsoft365defender_connector.py]:
```suggestion
    @staticmethod
    def _create_alert_artifacts(alert):
        return {
            'label': 'alert',
            'name': alert.get('title'),
            'source_data_identifier': alert.get('id'),
            'data': alert,
            'cef': alert
        }
```

File: microsoft365defender_connector.py [hunk 8]
```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
...
+
+    @staticmethod
+    def _create_incident_artifacts(incident):
+        return {'label': 'incident', 'name': 'incident Artifact', 'source_data_identifier': incident.get('id'),
```
Comment (bb-splunk) [inline microsoft365defender_connector.py]: I'd format this like the one above

**Review by bb-splunk — CHANGES_REQUESTED** [review]

File: microsoft365defender.json [hunk 9]
```diff
@@ -63,6 +84,42 @@
...
+                    "order": 0
+                },
+                "end_time": {
+                    "data_type": "numeric",
+                    "description": "Parameter ignored in this app",
+                    "order": 1
+                },
+                "container_count": {
+                    "data_type": "numeric",
+                    "description": "Parameter ignored for schedule/ interval polling only",
```
Comment (bbielinski-splunk) [inline microsoft365defender.json]:
```suggestion
                    "description": "Parameter ignored for schedule/interval polling only",
```

File: microsoft365defender_connector.py [hunk 10]
```diff
@@ -1190,6 +1350,11 @@ def initialize(self):
         self._tenant = config[DEFENDER_CONFIG_TENANT_ID]
         self._client_id = config[DEFENDER_CONFIG_CLIENT_ID]
         self._client_secret = config[DEFENDER_CONFIG_CLIENT_SECRET]
+        self._timeout = config.get('timeout', DEFAULT_TIMEOUT)
```
Comment (bbielinski-splunk) [inline microsoft365defender_connector.py]: Can we move "timeout" key to the consts to make it consistent with the rest of the keys above?

File: microsoft365defender_connector.py [hunk 11]
```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
...
+        Returns:
+            :return: status(phantom.APP_SUCCESS/phantom.APP_ERROR)
+        """
+        # Check that given time must not be before 1970-01-01T00:00:00Z.
+        if time < datetime.strptime("1970-01-01T00:00:00Z", DEFENDER_APP_DT_STR_FORMAT):
+            return phantom.APP_ERROR
+        return phantom.APP_SUCCESS
+
+    def _check_date_format(self, action_result, date):
+        time = None
```
Comment (bbielinski-splunk) [inline microsoft365defender_connector.py]: FYI: We don't need to initialize 'time' like that in this method

**Review by bbielinski-splunk — CHANGES_REQUESTED** [review]

File: microsoft365defender_connector.py (hunk 5 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: It was added when walrus operator was not added, need to remove this check

File: microsoft365defender_connector.py (hunk 8 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: actually, pycharm reformats it like this, however, I will update it

File: microsoft365defender.json (hunk 1 above)
Comment (aparekh-crest) [reply microsoft365defender.json]: Done

File: microsoft365defender.json (hunk 2 above)
Comment (aparekh-crest) [reply microsoft365defender.json]: changed

File: microsoft365defender.json (hunk 3 above)
Comment (aparekh-crest) [reply microsoft365defender.json]: Added

File: microsoft365defender_connector.py (hunk 4 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: Done

File: microsoft365defender_connector.py (hunk 5 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: Done

File: microsoft365defender_connector.py (hunk 6 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: Done

File: microsoft365defender_connector.py (hunk 8 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: Done

File: microsoft365defender_connector.py (hunk 10 above)
Comment (aparekh-crest) [reply microsoft365defender_connector.py]: Done

**Review by bbielinski-splunk — APPROVED** [review]

File: microsoft365defender_connector.py [hunk 12]
```diff
@@ -1150,6 +1150,167 @@ def _handle_update_alert(self, param):
...
         return action_result.set_status(phantom.APP_SUCCESS, DEFENDER_ALERT_UPDATED_SUCCESSFULLY)
 
+    @staticmethod
+    def _check_invalid_since_utc_time(time):
+        """Determine that given time is not before 1970-01-01T00:00:00Z.
+        Parameters:
+            :param time: object of time
+        Returns:
+            :return: bool(True/False)
+        """
```
Comment (bb-splunk) [inline microsoft365defender_connector.py]:
```suggestion
    def _check_invalid_since_utc_time(time: Datetime) -> bool:
        """Determine that given time is not before 1970-01-01T00:00:00Z."""
```

**Review by bb-splunk — APPROVED** [review]

**Review by dhwanis-crest — APPROVED** [review]

**Pull request merged**